xchange-stream-binance/src/main/java/info/bitrich/xchangestream/binance/BinanceStreamingMarketDataService.java

```diff
@@ -300,41 +325,204 @@ private void initRawOrderBookUpdatesSubscription(CurrencyPair currencyPair) {
+
+    @Override
+    public boolean isDisposed() {
+      return booksSubject.hasComplete() || booksSubject.hasThrowable();
+    }
+
+    private boolean isBookInitialised() {
+      return getBook() != null;
+    }
+
+    private OrderBook getBook() {
+      return bookRef.get();
+    }
+
+    private void bufferDelta(DepthBinanceWebSocketTransaction delta) {
+      synchronized (monitor) {
+        deltasBuffer.add(delta);
+      }
+    }
+
+    private Disposable asyncInitialiseOrderBookSnapshot() {
+      if (isBookInitialised()) {
+        LOG.info("Orderbook snapshot for {} was initialised before. Re-syncing.", currencyPair);
+
+        synchronized (monitor) {
+          bookRef.set(null);
+          deltasBuffer.clear();
+          bookLastUpdateId.set(0);
+        }
       }
+
+      return deltasObservable
+          .firstOrError()
+          .observeOn(bookSnapshotsScheduler)
+          .flatMap(delta -> fetchSingleBinanceOrderBookUpdatedAfter(delta))
+          .subscribe(
+              binanceBook -> {
+                final OrderBook book =
+                    BinanceMarketDataService.convertOrderBook(binanceBook, currencyPair);
+
+                synchronized (monitor) {
+                  bookRef.set(book);
+                  bookLastUpdateId.set(binanceBook.lastUpdateId);
+
+                  final List<DepthBinanceWebSocketTransaction> deltasPatchingSnapshot =
+                      deltasBuffer.stream()
+                          .filter(delta -> delta.getLastUpdateId() > binanceBook.lastUpdateId)
+                          .collect(Collectors.toList());
+
+                  deltasBuffer.clear();
+
+                  // Update the book with all buffered deltas (as probably nobody would like to be
+                  // notified with an already outdated snapshot).
+                  for (DepthBinanceWebSocketTransaction delta : deltasPatchingSnapshot) {
+                    if (!appendDelta(delta)) {
+                      disposables.add(asyncInitialiseOrderBookSnapshot());
+                    }
+                  }
+                }
+              },
+              error -> disposeWithError(error));
+    }
+
+    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
+    private boolean appendDelta(DepthBinanceWebSocketTransaction delta) {
+      synchronized (monitor) {
+        if (delta.getFirstUpdateId() > bookLastUpdateId.get() + 1) {
+          LOG.info(
+              "Orderbook snapshot for {} out of date (last={}, U={}, u={}).",
+              currencyPair,
+              bookLastUpdateId.get(),
+              delta.getFirstUpdateId(),
+              delta.getLastUpdateId());
+
+          return false;
+        } else {
+          bookLastUpdateId.set(delta.getLastUpdateId());
+
+          // FIXME
+          // The underlying impl is far from being optimal, LimitOrders could be created directly.
+          extractOrderBookUpdates(currencyPair, delta)
+              .forEach(update -> bookRef.get().update(update));
```

Comment:
why are you synchronizing here as well, you are using atomic data structures.

It seems like the issue is that you want all this stuff to get updated atomically together, I would make more sense to create a single object here and wrap it atomically updating it with `updateAndGet `or just use synchronize and don't use atomics.

Reply:
See the explanation added to previous comment. Just to add on top of that, the idea is to keep book snapshot and its lastUpdateId etc, synchronised, they need to be in sync to each other.

Reply:
I would prefer to have a minimum extra synchronization as possible because 1) it is not necessary 2) may be misleading or hard to understand why something unnecessary is there for future maintainers

Reply:
Also just to recap, I do understand the difference in synchronizations you have here (multiple operations needed to be treated atomically vs individual thread safe objects). But from a pure code perspective, we should only be doing what we need, everything else doesn't make sense.

Reply:
@earce applied Your suggestions, finally resigned from all atomic references and concurrent queue, left only the synchronised blocks, used to guard access to: book, bookLastUpdateId, deltasBuffer.

The last commit contains also a few related changes and fixed typos.